### ISSUE TITLE
Beta tester required?

### ISSUE THREAD

**Issue by Fabian188** [DrTimothyAldenDavis/SuiteSparse#159]
I very much appreciate the performance of cholmod!! Coupled with the availability!!! Thanks a lot for this and the active development!

We optionally use cholmod for our free academic software. 

I was happy to read that version 6.0 beta is cmake based and want to volunteer to beta-test. However there is no root CMakeLists.txt but a Makefile. 

Is the CMakeLists.txt just missing or is not intended to have a pure cmake variant?

**Comment by DrTimothyAldenDavis** [comment]
Thanks -- yes, it would be great for you to give it a try.

The lack of a top-level CMakeLists.txt is intentional.  Each package has its own CMakeLists.txt.  The top-level Makefile is very simple -- it just dives into each package and builds it with its own CMakeLists.txt.

This is because SuiteSparse is a kind of a la carte collection of packages, not a single package.

**Comment by Fabian188** [comment]
Well, it compiled flawless all in once. But can it be controlled to select what to be build?

**Comment by DrTimothyAldenDavis** [comment]
That would be a nice option to have, but I haven't written something to do that.  It can be done by going into each package, as in:

cd AMD/build ; cmake .. ; make ; sudo make install
cd ../../COLAMD/build ;  cmake .. ; make ; sudo make install

etc. (or use "make local ; make install" if you don't have sudo privileges).

Or, you can comment out individual lines in the top-level Makefile.

I suppose I could do this by adding additional make targets, so that "make cholmod" at the top-level SuiteSparse folder would build AMD, COLAMD, CAMD, CCOLAMD, SuiteSparse_config, etc, and then build CHOLMOD.

**Comment by Fabian188** [comment]
I made a mistake: I mixed up the current SuiteSparse-master directory I extracted from the downloaded zip with a SuiteSparse-master directory form March. In that one CHOLMOD, ... had no CMakeLists.txt yet. 

Sorry for the confusion. I first built the new code but was then always referring to the outdated code :(

**Comment by DrTimothyAldenDavis** [comment]
Yes, I need to use cmake 3.22 because I have to control the size of the BLAS integer, with BLA_SIZEOF_INTEGER.  That was added in cmake 3.22: https://cmake.org/cmake/help/latest/module/FindBLAS.html.  I hear you about not using such a new version but I really don't have another option.  This is in fact one reason why I couldn't use cmake extensively until recently.  Earlier versions of cmake didn't have the features I need.

I have to know which BLAS is in use when I build SuiteSparse, for several reasons:  (1) I have to know the BLAS integer size.  There is no standard way for BLAS libraries to provide both so that I can select them when doing the final link in the application (like "dgemm32" and "dgemm64").  Even the cblas interface is limited (it only allows for 32-bit BLAS, which is too limiting).  This is a serious issue with the BLAS API but I can't fix that.

(2) Some of my packages will be using non-portable functions inside the BLAS, if those are available, such as the Intel MKL function for telling the BLAS how many threads to use.  That will be controlled by #ifdef BLAS_Intel10_* (where * is any of the Intel BLAS options).  Similar thing for OpenBLAS.  Sadly, there is no standard way in the BLAS API for controlling this option, which is why many BLAS libraries have to add non-standard options.  Again, a serious issue with the BLAS API.  This feature is required for composing parallel algorithms (like CHOLMOD, GraphBLAS, ParU (to appear), and SPQR that call parallel BLAS.  I have multiply threads of mine that need to call the BLAS in parallel, and each BLAS call itself needs to be parallel.  That's a challenge, and to control it I need control to tell each BLAS call how many threads to use.

Both issues (1) and (2) are due to limitations in the BLAS and LAPACK interface, which I hope could be resolved in the future with a standard "cblas64_dgemm" kind of interface with int64_t integers, and a standard "cblas_set_num_threads" or something.  Until then, I will have to stick with the limitations of the API, which means I must know which BLAS and LAPACK libraries will be used when I build SuiteSparse.

I do have a single place where all SuiteSparse-wide policies are set. It's this file:
https://github.com/DrTimothyAldenDavis/SuiteSparse/blob/master/SuiteSparse_config/cmake_modules/SuiteSparsePolicy.cmake, a copy of which is also in GraphBLAS/cmake_modules because GraphBLAS doesn't rely on  SuiteSparse_config for historical reasons (I may change that in the future).

I have an NSTATIC option that allows you to turn off the build of the static libraries.

I wasn't aware of the option command in cmake.  What does that do any different than just:

`cd build ; cmake -DNSTATIC=1 .. ; make ; make install`

or with my top-level Makefile:

``CMAKE_OPTIONS="-DNSTATIC=1" make ; make install`

for example?  Is it a GUI thing where the option commands get imported into a CMake GUI, like MS Visual Studio?  (I don't use any cmake GUI).  I see that the command is limited to boolean variables.  That's a pity; some of my options need to be tokens like "Intel10_64lp" or "OpenBLAS", or strings like "52;60;80" for SUITESPARSE_CUDA_ARCHITECTURES.

**Comment by Fabian188** [comment]
Thank you for the explanation w.r.t. BLAS. Then I find it import to be documented how to enforce own blas (I have on my system apple Acceleration, openblas and mkl).

The "gui" is text based, just call ccmake instead of cmake (probably you need brew install extra-cmake-modules). It lets the -DVARIABLE=* stuff be done interactively. One also sees all options plus an overview of what is internally set (e.g. which blas is found). One needs to run configure (c) as often up to one can generate (g) the makefiles.

One can interactively set any value. With this selection from a list we set the blas for our application
```
set(USE_BLAS_LAPACK ${USE_BLAS_LAPACK_DEFAULT} CACHE STRING "BLAS/LAPACK implementation. May be one of OPENBLAS, MKL, NETLIB")
set_property(CACHE USE_BLAS_LAPACK PROPERTY STRINGS OPENBLAS MKL NETLIB)
```
The important point is the CACHE attribute in set().

**Comment by Fabian188** [comment]
This is ccmake instead of cmake for your CHOLMOD
<img width="866" alt="Bildschirmfoto 2022-11-08 um 13 00 02" src="https://user-images.githubusercontent.com/33481114/200558723-b2b1728b-424d-4dfd-952c-73bc16a9ede8.png">

**Comment by Fabian188** [comment]
With (t) I can see the hidden options and edit them (e.g. compiler flags)
<img width="1163" alt="Bildschirmfoto 2022-11-08 um 13 01 31" src="https://user-images.githubusercontent.com/33481114/200558939-01a8df9f-5a78-4e7a-a47c-bea652c36bb0.png">

**Comment by Fabian188** [comment]
Here is a last example. OpenBLAS is exactly how I know it with cmake:
```
cd OpenBLAS-0.3.21
mkdir build
cd build
ccmake ..
```
And then I can configure what I want. I my case, I would switch off BUILD_TESTING.

<img width="1170" alt="Bildschirmfoto 2022-11-08 um 13 37 37" src="https://user-images.githubusercontent.com/33481114/200566107-7f521049-c735-4c49-895c-8bbdd1436867.png">

**Comment by DrTimothyAldenDavis** [comment]
OK ... I've added all of this.  All my settings are not done via the "option" command, or "set ( ... CACHE ....) " for strings.  I'll post a new beta prerelease shortly.  ccmake should show you all these options, and they are also described in the README (see the dev2 branch, or the next beta coming out shortly).

**Comment by Fabian188** [comment]
I checked dev2, my intention was to build CHOLMOD and UMFPACK.
I did ccmake within build for SuiteSparse_config, AMD, COLAMD, CCOLAMD, CHOLMOD and UMFPACK.

I got the libs, which is what I need. So this works with cmake as before with make local - thanks!

I suggest to add the DEMO option and possibly have a root CMakeLists, which maybe just calls the other cmakes with the common settings as default.
